common/lib/xmodule/xmodule/util/adaptive_learning.py

```diff
@@ -0,0 +1,270 @@
+        try:
+            student = next(s for s in students if s.get('uid') == user_id)
+        except StopIteration:
+            student = None
+        return student
+
+    def get_students(self):
+        """
+        Return list of all students that external service knows about.
+        """
+        url = self.students_url
+        response = requests.get(url, headers=self.request_headers)
+        students = json.loads(response.content)
+        return students
+
+    def create_student(self, user_id):
+        """
+        Create student identified by `user_id` on external service,
+        and return it.
+        """
+        url = self.students_url
+        payload = {'uid': user_id}
+        response = requests.post(url, headers=self.request_headers, data=payload)
+        student = json.loads(response.content)
+        return student
+
+    def get_knowledge_node_student(self, block_id, user_id):
+        """
+        Return 'knowledge node student' object for user identified by `user_id`
+        and unit identified by `block_id`, or None if it does not exist.
+        """
+        # Get 'knowledge node student' objects
+        links = self.get_knowledge_node_students()
+        # Filter them by `block_id` and `user_id`
+        try:
+            link = next(
+                l for l in links if l.get('knowledge_node_uid') == block_id and l.get('student_uid') == user_id
+            )
+        except StopIteration:
+            link = None
+        return link
+
+    def get_knowledge_node_students(self):
+        """
+        Return list of all 'knowledge node student' objects for this course.
+        """
+        url = self.knowledge_node_students_url
+        response = requests.get(url, headers=self.request_headers)
+        links = json.loads(response.content)
+        return links
+
+    def create_knowledge_node_student(self, block_id, user_id):
+        """
+        Create 'knowledge node student' object that links student identified by `user_id`
+        to unit identified by `block_id`, and return it.
+        """
+        url = self.knowledge_node_students_url
+        payload = {'knowledge_node_uid': block_id, 'student_uid': user_id}
+        response = requests.post(url, headers=self.request_headers, data=payload)
+        knowledge_node_student = json.loads(response.content)
+        return knowledge_node_student
+
+    def create_knowledge_node_students(self, block_ids, user_id):
+        """
+        Create 'knowledge node student' objects that link student identified by `user_id`
+        to review questions identified by block IDs listed in `block_ids`, and return them.
+        """
+        knowledge_node_students = []
+        for block_id in block_ids:
+            knowledge_node_student = self.get_or_create_knowledge_node_student(block_id, user_id)
+            knowledge_node_students.append(knowledge_node_student)
+        log.error("KNOWLEDGE_NODE_STUDENTS: {}".format(knowledge_node_students))
```

Comment:
When I try to view the ACB in the LMS, this line is reporting a 500 error.  It tries 9 times to connect to the domoscio URL, and then outputs:
```
2016-12-28 03:54:13,813 ERROR 9304 [xmodule.util.adaptive_learning] adaptive_learning.py:238 - KNOWLEDGE_NODE_STUDENTS: [{u'status': u'500', u'error': u'Internal Server Error'}, {u'status': u'500', u'error': u'Internal Server Error'}]
```